downstairs/src/extent_inner_sqlite.rs

```diff
@@ -508,6 +505,77 @@ impl ExtentInner for SqliteInner {
 }
 
 impl SqliteInner {
+    /// Converts to a raw file for use with `RawInner`
+    ///
+    /// Returns the metadata and context slots, which should be positioned
+    /// directly after the raw block data in memory.
+    pub fn export_meta_and_context(
+        &mut self,
+    ) -> Result<Vec<u8>, CrucibleError> {
+        // Check whether we need to rehash.  This is theoretically represented
+        // by the `dirty` bit, but we're being _somewhat_ paranoid and manually
+        // forcing a rehash if any blocks have multiple contexts stored.
+        let ctxs = self.get_block_contexts(0, self.extent_size.value)?;
+        let needs_rehash = ctxs.iter().any(|c| c.len() > 1);
+
+        if needs_rehash {
+            // Clean up stale hashes.  After this is done, each block should
+            // have either 0 or 1 contexts.
+            self.fully_rehash_and_clean_all_stale_contexts(true)?;
+        }
+
+        let ctxs = self.get_block_contexts(0, self.extent_size.value)?;
+
+        use crate::{
+            extent::EXTENT_META_RAW,
+            extent_inner_raw::{
+                OnDiskDownstairsBlockContext, OnDiskMeta,
+                BLOCK_CONTEXT_SLOT_SIZE_BYTES, BLOCK_META_SIZE_BYTES,
+            },
+        };
+
+        // Record the metadata region, which will be right after raw block data
```

Comment:
Nit, I try not to use `extent` and `region` unless I'm specifically talking about Crucible's specific `extent` or `region`.  I might suggest "Record the metadata section, .." or another word of your choice here.

Reply:
Good idea, done in db9c6548cee97ae25d0bb5e6838bb0e74700ce44